### ISSUE TITLE
Intellisense priority issues breaks Go to definition

### ISSUE THREAD

**Issue by habovh** [clinyong/vscode-css-modules#63]
Hi,

I've been using this extension really fine on JS projects. It's great and really useful!

However on TS projects, it seems this extension is "fighting" for priority against the TS Intellisense provider. Unfortunately, the TS provider trumps this extension, and when using "Go to definition", it leads to the very generic type definition used to make the compiler happy instead of the actual class in the style module file.

This is rather annoying.

I've never developed a language server extension, but I was wondering if it was possible to increase priority of the suggestions/sources provided by this extension, so that even in case Intellisense matches 2 different definitions, the "chosen" one will always be the one from the extension?

As I've tried to work this out, I discovered that enabling the extension **after** the JS/TS language features are initialised, then the Go to definition works great. However when I quit and reopen VSCode, the default definition is reverted back to the generic TS one.

Note: I tried all of this without any TS plugin, using VSCode's bundled TS runtime, and on a Next.js project.

**Comment by iChenLei** [comment]
Yep, my friend(same company employee) encounted the same issue, but I have no idea to fix or improve ux. My friend's scenario is:
```ts
declare module '*.module.css' {
  const content: { [className: string]: string }
  export default content; 
}
```
A possible workaround is 
```diff
declare module '*.module.css' {
-  const content: { [className: string]: string }
+ const content any;
  export default content; 
}
```
> And then TS will give up type Intellisense. ( I checked it on my macbook

**Comment by habovh** [comment]
Ooh that's a clever trick! Makes sense that Intellisense would favour anything over `any`. Thanks for this hack, I wish it wasn't required however.

The "priority" is not great either with autocomplete. It's quite weird the extension is not able to communicate to VSCode that its autocompletion are 100% relevant and should be placed before random stuff.

**Comment by karlhorky** [comment]
Also just encountered this in a Next.js project (which has CSS modules types in the global types that come along with it in [`node_modules/next/types/global.d.ts`](https://github.com/vercel/next.js/blob/77f40ae9f1510ebb306d6c1563b989359b10190b/packages/next/types/global.d.ts#L21-L34)):

<img width="675" alt="Screenshot 2023-01-05 at 23 08 59" src="https://user-images.githubusercontent.com/1935696/210889854-6ca79246-3ca5-4f59-91be-4dbcbdaff5bc.png">

Altering the `global.d.ts` types file in `node_modules` to use `any` does avoid this problem, but I think requiring users to edit files in `node_modules` is probably not a good solution.

Is there any API in VS Code to remove / invalidate some other definition from within the extension? Or maybe even just re-defining the types to use `any` in the types for the extension?

**Comment by iChenLei** [comment]
> Is there any API in VS Code to remove / invalidate some other definition from within the extension?
-----------------
I don't know, I can't found any useful info from the [official api documentation](https://code.visualstudio.com/api/references/vscode-api) for this topic.

**Comment by karlhorky** [comment]
Hm, it seems like this may be "as designed", as mentioned here:

> The extension system is designed to be cooperative and so that a single extension cannot mute results of other extensions. We have no plans of changing that. However, if you think that the TypeScript/JavaScript extension behaves incorrect for a given scenario that please file a new, separate issue with steps to reproduce. Thanks

- https://github.com/microsoft/vscode/issues/76231#issuecomment-506597317

However, maybe this falls into "behaves incorrect for a given scenario", and deserves a bug in TypeScript, similar to this change to the "Duplicate Definitions" issue here:

- https://github.com/microsoft/TypeScript/issues/24861

@jrieken @mjbvz do you have any guidance here? It would be great to allow users to have the capability of opting out of [these extra definition entries](https://github.com/clinyong/vscode-css-modules/issues/63#issuecomment-1372851831) when a better one exists. This would enable one-click Go to Definition, which seems like the UX that is desirable for users. Ideally automatically by the extension, but it would also be acceptable to have the user manually opt out of specific definitions in some files (as long as they do not have to edit the `global.d.ts` file in their `node_modules`, which does not seem like a tenable solution).

**Comment by karlhorky** [comment]
@iChenLei another idea: since you're using [`languages.registerDefinitionProvider`](https://code.visualstudio.com/api/references/vscode-api#:~:text=registerDefinitionProvider(selector%3A%20DocumentSelector%2C%20provider%3A%20DefinitionProvider)%3A%20Disposable) here:

https://github.com/clinyong/vscode-css-modules/blob/4fc1c2aac1bffe88380656d54b99d6ac13151265/src/extension.ts#L23-L26

...what do you think about trying to also declare the same `declare module *.module.css/sass/scss` types like in the Next.js global config? (eg. with `any`, as below) Maybe then your type would override this?

```ts
declare module '*.module.css' {
  const classes: any
  export default classes
}

declare module '*.module.sass' {
  const classes: any
  export default classes
}

declare module '*.module.scss' {
  const classes: any
  export default classes
}
```

**Comment by zardoy** [comment]
@karlhorky Only in case of TypeScript/JavaScript/Vue there is an API to take full control over definitions / completions and so on returned by TS language service with TS plugins. I've made an **extension that removes generic definition**: https://github.com/zardoy/typescript-vscode-plugins/ (enabled by default with `miscDefinitionImprovement`, [source code](https://github.com/zardoy/typescript-vscode-plugins/blob/9252361ac3c2395f391afa14d548472ff6103c72/typescript/src/definitions.ts#L141)).

I also noticed that defition priority is not consistent. With `"editor.gotoLocation.multipleDefinitions": "gotoAndPeek"` it was always revealing `.css` file definition first, but this week it started to focus on `.d.ts` file instead, I have no idea why, but because of this I made solution mentioned above.

**Comment by karlhorky** [comment]
I'm using [the React CSS modules extension](https://marketplace.visualstudio.com/items?itemName=viijay-kr.react-ts-css) ([repo](https://github.com/Viijay-Kr/react-ts-css)) at the moment, which appears to avoid the problem somehow... 🤔

**Comment by karlhorky** [comment]
@zardoy I'm not sure what you mean by "TypeScript/JavaScript/Vue", but reading the source code of your extension [TypeScript Essential Plugins](https://marketplace.visualstudio.com/items?itemName=zardoy.ts-essential-plugins) ([repo](https://github.com/zardoy/typescript-vscode-plugins)) it seems there may be code that could provide a way forward. Some interesting parts:

```ts
    proxy.getDefinitionAndBoundSpan = (fileName, position) => {
      const prior = info.languageService.getDefinitionAndBoundSpan(fileName, position)
      // ...
      prior.definitions = prior.definitions.filter(({ fileName, containerName, containerKind, kind, name, ...rest }) => {
        // ...
        if (moduleDeclaration?.name.text === '*.module.css') return false
        return true
      })
      // ...
    })
```

- https://github.com/zardoy/typescript-vscode-plugins/blob/a102a76786a0292aa8a9a46fb6cc43595e8871a7/typescript/src/definitions.ts#L7-L8
- https://github.com/zardoy/typescript-vscode-plugins/blob/a102a76786a0292aa8a9a46fb6cc43595e8871a7/typescript/src/definitions.ts#L123
- https://github.com/zardoy/typescript-vscode-plugins/blob/a102a76786a0292aa8a9a46fb6cc43595e8871a7/typescript/src/definitions.ts#L141


If I'm understanding correctly, this is filtering out previous definitions if they have the name `*.module.css`, which could also be expanded to include `*.module.scss` and `*.module.sass`.

@clinyong what do you think about this approach?

**Comment by zardoy** [comment]
> I'm not sure what you mean by "TypeScript/JavaScript/Vue"

@karlhorky I meant that it is possible to fully control language server features of these languages.

> If I'm understanding correctly, this is filtering out previous definitions

Yes, the logic is quite simple, however I included additional pattern check to ensure we don't do extra work on every definition. And yes it can be extended, I just forgot to add .scss and sass extensions, will fix it.

Also feel free to ask whatever you want!

**Comment by Stan-Stani** [comment]
For me, the weird thing is even if I replace Next.js types for scss module imports with `any` I just get `any` as the hover type definition. But if I control click it takes me to the appropriate selector in `styles.module.scss`. Any one have any tips to get hover definitions working?

**Comment by alan-albuquerque** [comment]
Could this be solved if we have an option to exclude folders or files by a glob pattern?
```
  "cssModules.ignorePaths": [
    "**/node_modules/**",
  ],
```
